lib/partitionAdapter.ts

```diff
@@ -0,0 +1,45 @@
+/* Copyright (c) 2024. SailPoint Technologies, Inc. All rights reserved. */
+
+import { Context, readConfig, Response, StdAccountListInput, StdAccountListOutput } from ".";
+import { Partition } from "./partition";
+
+export interface PartitionAdapter {
+	getPartitions(context: Context, input: StdAccountListInput): Promise<Array<Partition>>
+	list(context: Context, input: StdAccountListInput, res: Response<StdAccountListOutput>, partition?: Partition): Promise<void>
+}
+
+export class PartitionHandler {
+	public partitionAdapter;
+
+	constructor(partitionAdapter: PartitionAdapter) {
+		this.partitionAdapter = partitionAdapter;
+	}
+
+	/**
+ 	 * Handles partitioning operation for the connector which supports the partitioning.
+	 * 
+	 * @param context Context
+	 * @param input StdAccountListInput
+	 * @param res Response<StdAccountListOutput>
+ 	 */
+	async handler(context: Context, input: StdAccountListInput, res: Response<StdAccountListOutput>): Promise<void> {
+		const config = await readConfig();
+		const isPatitionEnabled = config.partitionAggregationEnabled;
+		var partitionList: Array<Partition> | undefined;
+
+		// Get the partitions from the connector if partitining is enabled
+		if (isPatitionEnabled) {
+			partitionList = await this.partitionAdapter.getPartitions(context, input);
+		}
+
+		if (!partitionList) {
+			let promises = [];
+			for (const partition of partitionList!) {
+				promises.push(this.partitionAdapter.list(context, input, res, partition));
+			}
+			await Promise.all(promises);
+		} else {
+			return await this.partitionAdapter.list(context, input, res, undefined);
+		}
```

Comment:
Can we add logging here to allow troubleshooters to determine if:

1. We used partition aggregation or not.
2. How long retrieving the partitions took.
3. The number of partitions returned.
4. Might also be useful to log when we finishing aggregating a specific partition.

Reply:
Point 1-3 addressed.
Point 4 - Shouldn't this get logged at the connector level? I think we do not have control to check when a particular partition is finishing.

Reply:
For 4 I was thinking we could do something like:

```typescript
let partitionsDone = 0
for (const partition of partitionList!) {
  promises.push(async () => {
    let startTime = new Date().getTime();
    await this.partitionAdapter.list(context, input, res, partition);
    partitionsDone += 1
    this.childLogger.info(`Partition ${} of ${} done after ${}`)
  })
}
await Promise.all()
```

The reason I think this is valuable at the library level is because it gives us a good idea of how good of a job the connector is doing at creating equally sized partitions. I remember seeing a case of partitioning on java connectors where one of the partitions ended up being much larger than the others so we weren't taking advantage of as much parallelism as we could have.

Reply:
Makes sense. Update the code.